pkg/policy/policy_test.go

```diff
@@ -518,6 +534,107 @@ func TestPolicies_Eval(t *testing.T) {
 			policies:           make(map[string]string),
 			expectedError:      policy.PoliciesNotFoundError,
 		},
+		{
+			name:               "using a custom rule - check passed",
+			resource:           simpleNginxPod,
+			useBuiltInPolicies: false,
+			policies: map[string]string{
+				"policy.uses_image_tag_latest.kinds": "Pod",
+				"policy.uses_image_tag_latest.rego": `
+   package trivyoperator.policy.k8s.custom
+   __rego_metadata__ := {
+        "id": "CUSTOMCHECK",
+        "title": "custom check title",
+        "severity": "LOW",
+        "type": "Kubernetes Security Check",
+        "description": "custom check description",
+    }
+
+   alwaysFalse {
+      1 == 0
+   }
+
+   deny[res] {
+        alwaysFalse
+		res := {
+				"msg": "the check should always pass",
+			}
+   }`,
+			},
+			results: []Result{
+				{
+					Metadata: Metadata{
+						ID:          "CUSTOMCHECK",
+						Title:       "custom check title",
+						Description: "custom check description",
+						Severity:    "LOW",
+						Type:        "Kubernetes Security Check",
+					},
+					Success: true,
+				},
+			},
+		},
+		{
+			name:               "using a custom rule - check failed",
+			resource:           simpleNginxPod,
+			useBuiltInPolicies: false,
+			policies: map[string]string{
+				"policy.uses_image_tag_latest.kinds": "Pod",
+				"policy.uses_image_tag_latest.rego": `
+   package trivyoperator.policy.k8s.custom
+   __rego_metadata__ := {
+        "id": "CUSTOMCHECK",
+        "title": "custom check title",
+        "severity": "LOW",
+        "type": "Kubernetes Security Check",
+        "description": "custom check description",
+    }
+
+   alwaysTrue {
+      1 == 1
+   }
+
+   deny[res] {
+        alwaysTrue
+		res := {
+				"msg": "the check should be always failed",
+			}
+   }`,
+			},
+			results: []Result{
+				{
+					Metadata: Metadata{
+						ID:          "CUSTOMCHECK",
+						Title:       "custom check title",
+						Description: "custom check description",
+						Severity:    "LOW",
+						Type:        "Kubernetes Security Check",
+					},
+					Messages: []string{"the check should be always failed"},
+					Success:  false,
+				},
+			},
+		},
+		{
+			name:               "using a custom rule with incorrect name",
+			resource:           simpleNginxPod,
+			useBuiltInPolicies: false,
+			policies: map[string]string{
+				"policy.uses_image_tag_latest.kinds": "Pod",
+				"policy.uses_image_tag_latest.rego": `
+   package myfunnyname.policy.k8s.custom
+   __rego_metadata__ := {
+        "id": "CUSTOMCHECK",
+        "title": "custom check title",
+        "severity": "LOW",
+        "type": "Kubernetes Security Check",
+        "description": "custom check description",
+    }
+
+   deny[{"msg": "this message should be hidden"}]`,
+			},
+			expectedError: "failed to run policy checks on resources",
+		},
```

Comment:
I don't think this testcase is correct because the behavior today is wrong (IMO). I say this because if the check does not apply to a resource, it should not return either a result nor an error. But today we return an error in such a scenario https://github.com/aquasecurity/trivy-operator/blob/b457d371f57ca7ae9a82ce41756c2092f5b6c252/pkg/policy/policy.go#L253-L255

I think we should remove this error return as we should simply return no results and no error. This is possible due to the fact that a check may not apply to a resource.

Reply:
@simar7 Actually, I agree with you, but I'm unsure about my understanding policy checks.

I couldn't reproduce a case when `scanResult` is `nil` for correct policies.
there is a test when I try to apply POD policy to RBAC resource: https://github.com/afdesk/trivy-operator/pull/1/files
`scanResult` contains metadata from the check, but this one isn't passed.

> a check may not apply to a resource.

it seems it's not possible by design.
there is a check before evaluating policies to resources.
https://github.com/afdesk/trivy-operator/blob/9b6b5f463ab2b22e9be214eac43d3a83a7bc6f10/pkg/configauditreport/controller/resource.go#L181-L189

maybe I missed something here

Reply:
As we discussed offline - the only cases where scanResult & err both can be nil are if: 1. setup of checks fails, 2. namespace mismatch. 3. kind mismatch

Since we know 3. cannot happen at this stage (subtype filtering of kind is already performed) and 1. isn't the case at this stage, that would leave us with only 2. as the potential reason.

After testing this further, I could confirm that it is indeed the case as this logic is the one that takes place https://github.com/aquasecurity/trivy/blob/main/pkg/iac/rego/scanner.go#L194-L197